src/Data/Nat/Binary/Minus.agda

```diff
@@ -0,0 +1,364 @@
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Nat.Binary.Minus where
+
+open import Algebra using (Magma)
+import Algebra.FunctionProperties as FuncProp
+open import Data.Empty using (⊥)
+open import Data.Fin using (#_)
+open import Data.Nat as ℕ using (ℕ)
+open import Data.Nat.Binary.Base
+open import Data.Nat.Binary.Properties
+import Data.Nat.Properties as ℕₚ
+open import Data.Product using (_×_; _,_; proj₁; proj₂; ∃)
+open import Data.Sum using (inj₁; inj₂)
+open import Data.Vec using ([]; _∷_)
+open import Function using (_∘_; _$_)
+open import Level using (0ℓ)
+open import Relation.Binary using (Tri; _Preserves_⟶_; _Preserves₂_⟶_⟶_)
+open import Relation.Binary.PropositionalEquality as P using
+  (_≡_; _≢_; _≗_; refl; sym; trans; cong; cong₂; resp₂; subst; magma;
+   module ≡-Reasoning)
+open import Relation.Nullary using (Dec; yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+open import Algebra.FunctionProperties {A = ℕᵇ} _≡_
+open import Algebra.Properties.CommutativeSemigroup +-commutativeSemigroup
+  using (xy∙z≈y∙xz; x∙yz≈y∙xz)
+open import Algebra.Solver.CommutativeMonoid +-0-commutativeMonoid
+
+private module FP-Bin = FuncProp (_≡_ {A = ℕᵇ})
+open Tri
+
+infixl 6 _∸_
+
+_∸_ : Op₂ ℕᵇ
+zero     ∸ _        = 0ᵇ
+x        ∸ zero     = x
+2[1+ x ] ∸ 2[1+ y ] = double (x ∸ y)
+1+[2 x ] ∸ 1+[2 y ] = double (x ∸ y)
+2[1+ x ] ∸ 1+[2 y ] with x <? y
+... | yes _ = 0ᵇ
+... | no _  = 1+[2 (x ∸ y) ]
+1+[2 x ] ∸ 2[1+ y ] with x ≤? y
+... | yes _ = 0ᵇ
+... | no _  = pred (double (x ∸ y))
+
+x∸0≡x : ∀ x → x ∸ 0ᵇ ≡ x
+x∸0≡x zero     = refl
+x∸0≡x 2[1+ _ ] = refl
+x∸0≡x 1+[2 _ ] = refl
+
+even∸odd-for≥ : ∀ {x y} → x ≥ y → 2[1+ x ] ∸ 1+[2 y ] ≡ 1+[2 (x ∸ y) ]
+even∸odd-for≥ {x} {y} x≥y  with x <? y
+... | no _    = refl
+... | yes x<y = contradiction x≥y (<⇒≱ x<y)
+
+odd∸even-for> : ∀ {x y} → x > y → 1+[2 x ] ∸ 2[1+ y ] ≡ pred (double (x ∸ y))
+odd∸even-for> {x} {y} x>y  with x ≤? y
+... | no _    = refl
+... | yes x≤y = contradiction x>y (≤⇒≯ x≤y)
+
+x≤y⇒x∸y≡0 :  ∀ {x y} → x ≤ y → x ∸ y ≡ 0ᵇ
+x≤y⇒x∸y≡0 {zero} {_} _           =  refl
+x≤y⇒x∸y≡0 (inj₁ (even<even x<y)) =  x≡0⇒double[x]≡0 (x≤y⇒x∸y≡0 (inj₁ x<y))
+x≤y⇒x∸y≡0 {2[1+ x ]} {2[1+ .x ]} (inj₂ refl) =
+  x≡0⇒double[x]≡0 (x≤y⇒x∸y≡0 {x} {x} (inj₂ refl))
+x≤y⇒x∸y≡0 {2[1+ x ]} {1+[2 y ]} (inj₁ (even<odd x<y))  with x <? y
+... | yes _  =  refl
+... | no x≮y =  contradiction x<y x≮y
+x≤y⇒x∸y≡0 {1+[2 x ]} {2[1+ y ]} (inj₁ (odd<even x≤y)) with x ≤? y
+... | yes _  =  refl
+... | no x≰y =  contradiction x≤y x≰y
+x≤y⇒x∸y≡0 {1+[2 _ ]} {2[1+ _ ]} (inj₂ ())
+x≤y⇒x∸y≡0 {1+[2 _ ]} {1+[2 _ ]} (inj₁ (odd<odd x<y)) =
+  x≡0⇒double[x]≡0 (x≤y⇒x∸y≡0 (inj₁ x<y))
+x≤y⇒x∸y≡0 {1+[2 x ]} {1+[2 .x ]} (inj₂ refl) =
+  x≡0⇒double[x]≡0 (x≤y⇒x∸y≡0 (≤-refl {x}))
+
+x∸x≡0 : ∀ x → x ∸ x ≡ 0ᵇ
+x∸x≡0 x =  x≤y⇒x∸y≡0 (≤-refl {x})
+
+x>y⇒x∸y≢0 : ∀ {x y} → x > y → x ∸ y ≢ 0ᵇ
+x>y⇒x∸y≢0 {zero} {_} ()
+x>y⇒x∸y≢0 {x'@(2[1+ x ])} {zero} _ =  subst (_≢ 0ᵇ) x'≡x'∸0 2[1+x]≢0
+  where
+  x'≡x'∸0 = sym (x∸0≡x x')
```

Comment:
This proof is short enough that you don't need to inline it. Alternatively you could use the `rewrite` syntax.

Reply:
Changed to ``... = subst (_≢ 0ᵇ) (sym (x∸0≡x x')) 2[1+x]≢0``